### ISSUE TITLE
doc2vec / Paragraph Vectors Uses word2vec loss function

### ISSUE THREAD

**Issue by rothn** [deeplearning4j/deeplearning4j#4942]
#### Issue Description

* In the current DL4J implementation, paragraph vectors does not use the doc2vec loss function
* In fact, it is trained in what seems to be exactly the same way as word2vec
* Specifically, in this paper: https://cs.stanford.edu/~quocle/paragraph_vector.pdf
    * The parameter "h" should be constructed from a mathematical concatenation (or average, as DL4J uses) of the context words in W and the paragraph vector from D
    * This is mentioned in the second paragraph of section 2.2 and clarified in the third paragraph
    * Currently, DL4J uses equation 1 without modification
 
This results in an improperly trained model. In many cases, users may simply be better off by averaging word vectors instead of inferring a paragraph vector since the word vectors were not trained w.r.t. a paragraph vector and thus the result of a prediction task could be arbitrary.

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j 0.9.1
* All platforms

**Comment by raver119** [comment]
I need to check that. As I remember impl was based on some python implementation.

Thanks for highlighting.

With best regards,
Vyacheslav K.

**Comment by rothn** [comment]
If you're going back into the implementation, it may also be worth noting
that mathematical concatenation has been shown to work better than an
average.

Also, I really like this implementation (in Python) from the Machine
Learning Cookbook:
https://github.com/PacktPublishing/TensorFlow-Machine-Learning-Cookbook/blob/master/Chapter%2007/doc2vec.py

In this case, the lines that DL4J seems to be missing are 109 and 112

**Comment by rothn** [comment]
?

**Comment by raver119** [comment]
Sorry for the delay, forgot to post answer here :(

You're fererring to PV-DM, which is basically CBOW.

Here's the piece of code where label gets sent to cpp together with context words for averaging:
https://github.com/deeplearning4j/deeplearning4j/blob/0be227a98302585ccb229e5275967f947a1a3a3e/deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/embeddings/learning/impl/sequence/DM.java#L128-L161

And here's the piece of code where actual averaging happens:
https://github.com/deeplearning4j/deeplearning4j/blob/38626210a2bf61d06bc14123f0121cacba4752fa/libnd4j/include/ops/declarable/helpers/cpu/sg_cb.cpp#L547-L570

And it was like that since day 0.

Idea of concatenation is nice, and we'll consider adding that as option.

Once again, sorry for late response.
